Reserve only caml_minor_heap_max_wsz * Max_domains for the minor heap

sabine opened the pull request:
This addresses https://github.com/ocaml/ocaml/issues/10971 in the sense that
* we do no longer reserve a huge, fixed-size-per-domain amount of memory for the minor heaps, instead the initial minor heap reservation is `s * Max_domains`
* domains can use `Gc.set` to increase their minor heap size beyond the reservation (causes STW) or modify their minor heap size within the current reservation (doesn't cause STW)
* the size of the reservation is always `caml_minor_heap_max_wsz * Max_domains`

It's now possible to use tools like AFL and Valgrind. The changes to `Gc.set` do not break existing code.

There is the possibility of a different pull request to add an OCAMLRUNPARAM option to set `Max_domains`, but we intentionally did not include this here to keep the patch minimal.

## summary
When a domain uses Gc.set to increase its own minor heap size to be larger
than the current reservation, we reserve a new memory area for
all minor heaps. This is done inside a STW section:

1. every domain frees its own heap,
2. inside a global barrier, the final domain unreserves the old minor heap
area, then reserves the new minor heap area. This domain also adjusts
the minor heap area pointers of all domains while the other domains
wait for the end of the barrier,
3. every domain allocates its own heap with size `Caml_state->minor_heap_wsz`.

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

xavierleroy commented:
Looks nice, thanks!  However the tests fail because you have a bunch of C functions that are not prefixed by `caml_` and should therefore be declared `static`.  Can you please sort this out?

gasche commented:
This PR contains 20% of the feature set discussed in #10971 and 80% of the difficulty: we don't handle changing the maximum number of domains here, but we still have to un-map and re-map minor heaps at runtime, which is the hard part. We hope to be able to add other features easily on top. (@sabine and myself have prototyped those additional features, but we went back to the bare minimum when we had to track a segfault by staring inquisitively at the diff.)

In the middle of a Stop-The-World (STW) section, we unmap and then re-map the minor heap space, and we modify the `dom_internal` structure of *all* domains (whether they are registered as STW participants or not) to store the new boundaries  of each domain reserved space (`domain_self->minor_heap_area`, `domain_self->minor_heap_area_end`). The fact that this works is not obvious, it relies on STW synchronization guarantees that are currently being documented in #11072.

When @sabine and myself wrote the code of the present PR, I was assuming that no code touching the domain structures could not run in parallel with STW sections (before a STW barrier), and that this justified the correctness of our change. In fact the assumption, as discussed in #11072, is not entirely correct: domain initialization code cannot run in parallel with STW sections, but domain cleanup code can.
Fortunately, neither the domain initialization nor the domain cleanup code currently mutate the minor heap reservation or its boundaries `domain_self->minor_heap_area{,_end}`. The variables are read (not mutated) if a domain decides to change its actual (committed) minor heap size, which only occurs through `Gc.set` in mutator code that cannot run in parallel with STW sections.

All this to say: reviews are warmly welcome!

(cc @Engil, who did a pre-review already, and @ctk21 who is most probably interested in looking at this.)

abbysmal commented:
Sorry for the delay in reviewing this: I will be taking a look at this PR very shortly, at the very least by the end of Monday I will have a review for it.

Thank you very much for this work!

ctk21 commented on runtime/gc_ctrl.c:
```diff
@@ -197,13 +199,31 @@ CAMLprim value caml_gc_set(value v)
+                 Caml_state->minor_heap_wsz / 1024);
+    caml_gc_log ("set minor heap size: %"
+                 ARCH_SIZET_PRINTF_FORMAT "uk words", newminwsz / 1024);
+    /* FIXME: when (newminwsz > caml_minor_heap_max_wsz) and
```
Is there a reason not to move the functionality of `caml_update_minor_heap_max` inside `caml_set_minor_heap_size`?
Wouldn't that deal with the `FIXME` and also reduce the surface of functions to just `caml_set_minor_heap_size`.

abbysmal replied:
Agreed, there's no other entrypoint to `caml_set_minor_heap_size` (nor to `caml_update_minor_heap_max`) so this wouldn't add any unwanted feature to either functions.

gasche replied:
Moving the functionality is a good suggestion, but I believe that it would not make the FIXME disappear.

When we update the minor heap max, we need to change the minor heaps area and then each domain then re-commits its own minor-heap memory -- towards the end of the STW section. In particular, the current domain (which is planning to change its minor heap size) recommits its "previous" minor heap size. And then we run `caml_set_minor_heap_size` which will decommit and recommit again; there is a duplication here (the domain decommits and recommits twice) which is the FIXME.

If we simply move `caml_set_minor_heap_size` into `caml_update_minor_heap_max`, the duplciation will still be present. If we wanted to fix the FIXME, we would need to pass the information about the desired *new* heap size to the STW section. But this would use a sensibly more complex API: we need to tell each domain which size to use (with most domains not changing their own size... except if we later implement a different "change the minor heap size of all domains" knob). I think it is wiser to wait for the "change the maximum number of domains" feature set to land (as further PRs), and see what API we have at this point, to revisit this question.

Note: one thing we tried is to update `Caml_state->minor_heap_wsz` before calling `caml_update_minor_heap_max`, so that when the domain recommits its minor heap, it naturally uses the new setting. But I don't trust this way of doing things, because it means that between the update and the recommit the domain is in a slightly inconsistent state, with a `minor_heap_wsz` setting that is inconsistent with its actual `minor_young_end - minor_young_start` span. This could work if we audit the code carefully to ensure that we never rely on `minor_heap_wsz` (except when committing new memory), but the code running between the update and the recommit contains a lot of subtle GC functions and I'm not sure about adding new delicate constraints to them. We could also consider having two values in the domain state, "current minor heap size" and "desired minor heap size", and have the recommit use the "desired" information.

I'm not sure that the complexity of dealing with this FIXME is worth it. Maybe the best choice is to keep the code somewhat naive, with the redundant decommit/recommit, knowing that this operation is very rare anyway.

sabine replied:
I think that moving `caml_update_minor_heap_max` into `caml_set_minor_heap_size` could work:

If we can easily identify the domain that initiated the STW section (which, if I understand correctly, we can, since it is `participating[0]`), our protocol would be:

In `caml_set_minor_heap_size`, the domain frees its own heap before initiating the STW section.

During the STW section,

1. all domains except `participating[0]` free their minor heap.
2. As previously, one domain modifies all domains' minor heap areas during the STW section.
3. all domains except `participating[0]` allocate their minor heap.

After the STW section, back in `caml_set_minor_heap_size`, the domain allocates its own own heap in the desired new size.

Does that make sense? I'll implement and see.

gasche replied:
(I think it is not `participating[0]`, which is likely to be domain-0, but rather `stw_leader`, but I'm not sure.)

I think we want to use the same STW code to deal with Max_domains update in the near future. In the case where we want to update Max_domains, we want a different protocol because the current domain does not play a special role. Similarly, if we implement a "resize the minor heap for all active domains" entry point in Gc, we need a different interface again. Presonally I would be in favor of replacing the "Max_domains is a constant" part, which is a more severe limitation than this FIXME, before we look at what the STW interface is and whether we want to add more stuff to optionally change the leader's minor heap size.

sabine replied:
You're right, it's not `participating[0]`.

For max_domains, I am not 100% sure that changing this limit at runtime has a use case that warrants taking on the additional implementation complexity. It looks to me like

a. keeping the default number of max_domains at 128 (which will work for the vast majority of architectures), and
b. introducing an OCAMLRUNPARAM to set max_domains

is a reasonable alternative that would only require us to change the initialization code of the domain-indexed data structures.

Still, implementing the ability to change max_domains via `Gc.set` seems interesting in terms of identifying invariants and current limitations of some parts of the runtime.

sabine replied:
Identifying the domain who initiated the STW section via `((dom_internal *) atomic_load_acq(&stw_leader))->state`, and making this domain responsible for allocating its own heap in `caml_set_minor_heap_size`, after the STW section, seems to work: the tests pass. (https://github.com/sabine/ocaml/commit/f37be6ba8b56e78a60f052199f13d4748ea9352d)

One open question: if we do make max_domains changeable at runtime, should that live on `Gc.set` or on the `Domain` interface?

sabine replied:
In the change_minor_heap_size_FIXME branch https://github.com/sabine/ocaml/commit/f37be6ba8b56e78a60f052199f13d4748ea9352d:

The domain on which `Gc.set` was called initiates the STW section if the requested size is bigger than the current reservation.

During the STW section,

    all domains free their minor heap.
    As previously, one domain modifies all domains' minor heap areas during the STW section.
    all domains except the initiating domain (stw_leader) allocate their minor heap.

After the STW section, back in caml_set_minor_heap_size, the STW-initiating domain allocates its own own heap in the desired new size.

sabine replied:
Can we leave this FIXME in the current PR? Doing it like this, with the spurious reallocation of the leader domain's minor heap was the least invasive change. This patch does the heavy lifting of introducing a STW section to deallocate/unreserve/reserve/allocate the minor heaps.

A follow-up PR that resolves the FIXME by adjusting the semantics of `Gc.set` to change the minor heap size of all domains (with resulting simplifications in the codebase) is mostly ready.

ctk21 reviewed (commented):
Good feature that fixes some real problems with large virtual allocations. I believe the code to work as intended. 

I was surprised that a new STW section was introduced rather than just adding the functionality onto the end of the existing minor heap collection; you always have to empty the minor heaps to resize. However the implementation here works fine.

I see one possibility that might simplify things and deals with the `FIXME` in the code: I need convincing that we really need to expose both `caml_set_minor_heap_size` and `caml_update_minor_heap_max`.

abbysmal commented on runtime/caml/eventlog.h:
```diff
@@ -84,7 +84,8 @@ typedef enum {
     EV_MINOR_LOCAL_ROOTS_PROMOTE,
-    EV_DOMAIN_CONDITION_WAIT
+    EV_DOMAIN_CONDITION_WAIT,
+    EV_DOMAIN_RESIZE_HEAP_RESERVATION,
```
After the eventring PR is merged this likely won't quite matter but for completeness sake, this change should be replicated in `tools/eventlog_metadata.in`.

kayceesrk replied:
@Engil would you be able to add an issue to track this?

sabine replied:
I opened an issue at https://github.com/ocaml/ocaml/issues/11125 and added the missing info in `tools/eventlog_metadata.in`

abbysmal commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
-  void* heaps_base;
+/* minor heap initialization and resizing */
 
+static uintnat get_minor_heap_reservation_size() {
```
Maybe better to explicit this name's unit (`size` -> `bsize`)?
Not sure if we usually tend to consider that `size` by default is byte and `wsize` is the exception (designating words).
I think disambiguation is good on function names nonetheless.
(avoid having to go to the definition and checks for the `Bsize_wsize` call to make sure the math checks out.)

sabine replied:
Agreed. This is a clear improvement in readability, and possibly a thing that we should be doing everywhere to make it easier to spot mistakes and easier for newcomers to understand the codebase.

abbysmal reviewed (commented):
Thank you very much for this PR @sabine, @gasche !
I already did a preview a few weeks back with @gasche, and the changes looks good to me.
I have a few minor nitpicks; however I think the STW segment is pretty straightforward and has no flaws I can see.

As for the whether or not adding a new STW segment is preferable to reusing one: I tend to be a neutral about the notion. (at least here.)
We made similar changes when dealing with DLABs and I was not necessarily fond of extending an already complex STW segment (minor collection) to support resizing/slabbing of the minor segments.

I think the use case here is fine, and the general world of advice for dealing with STW segments should be to avoid touching or adding them as much as possible. (which I think @ctk21 has stated many times in the past :-))

Food for thoughts as well would be could this segment be reused for later extension of the minor heap allocator (the return of slabbing?).
Or would there be a mix of things between this new STW segment and the regular minor collection segment?
I think if we were to have a fancier, more dynamic allocator for the minor heap, if STW is required, reusing this segment rather than using the minor heap collection one should be fine.

Lastly, another item to ponder upon one day: I see @sabine and @gasche are also now enjoying the virtue of the burn-style tests! (a true Multicore staple.)
I wonder if we should one day factor some common bits in the ones we have, if only to have a streamlined control interface.
(which is a throwback to some discussions around the testsuite and parallel testcases)

kayceesrk commented on runtime/domain.c:
```diff
@@ -373,15 +392,26 @@ asize_t caml_norm_minor_heap_size (intnat wsize)
-    [domain_state->young_end].
+     [domain_state->young_start]
+   and
+     [domain_state->young_end].
```
For a given domain, do

1. `assert (young_start == minor_heap_area)`
2. `assert (young_end == minor_heap_area + minor_heap_wsz * sizeof(value))`
3. `assert (minor_heap_area_end == caml_minor_heap_max_wsz * sizeof(value))`

always hold? 

IINM I don't think we modify `young_start` and `young_end` except at those places where we modify `minor_heap_area` and `minor_heap_wsz`. If so,

1. Can we document these assertions here?
2. Perhaps we can get rid of `minor_heap_area` and `minor_heap_area_end` in a separate PR? They were necessary in the early versions of multicore OCaml where the domain state for a given domain was placed just before `young_start`. We did not have a dedicated register for domain state. We used the allocation pointer to find the domain state. The `minor_heap_area*` variables don't seem necessary now.

gasche replied:
Thanks for the perspective here.

As I documented in those comments, my mental model is that `minor_heap_area{,_end}` is about the "reserved" area for the domain minor heap, while `young_{start,end}` is about the *committed* memory for the minor heap. So no, I think that these assertions do not hold in general, `young_end` is typically sensibly smaller than `minor_heap_area_end`. (For example: a domain may have "reserved" 2Gio of address space, but only uses a few kilobytes or megabytes for its minor heap.)

Currently `minor_heap_area{,_end}` can be recomputed easily from the domain id and `minor_heap_wsz`, because the placement scheme is very regular, so we could still do without those variables. But having them gives us an easy way to experiment with other schemes if we want to in the future. (For example, we could consider having minor heaps of different reserved sizes for each domain, or not-reusing the minor heap area of a previous domain with the same id for locality reasons.)

Finally: in the transient state where we are initializing a domain, or uninitializing a domain, or changing the minor heaps area, we can observe `young_start == young_end == NULL` while `minor_heap_area{,_end}` is already set. This is another case where the assertions/invariant you propose are not satisfied, but of course this is very temporary.

kayceesrk replied:
That makes sense. Thanks!

kayceesrk commented on runtime/domain.c:
```diff
@@ -398,8 +438,25 @@ int caml_reallocate_minor_heap(asize_t wsize)
       (void*)domain_self->minor_heap_area,
       (char*)domain_state->young_end - (char*)domain_state->young_start);
```
Reading this comment https://github.com/ocaml/ocaml/pull/11082/files#diff-67115925103982a8ebeb085cfab5ef31a182c9a442bc51e053934364d3750dafR383-R387: 

```suggestion
      (char*)domain_state->minor_heap_wsz);
```

kayceesrk commented on runtime/domain.c:
```diff
@@ -373,15 +392,26 @@ asize_t caml_norm_minor_heap_size (intnat wsize)
-int caml_reallocate_minor_heap(asize_t wsize)
-{
+
+void caml_check_minor_heap() {
```
This function is not used outside this file. Inline it.

```suggestion
Caml_inline void check_minor_heap () {
```

kayceesrk commented on runtime/domain.c:
```diff
@@ -398,8 +438,25 @@ int caml_reallocate_minor_heap(asize_t wsize)
+
   wsize = caml_norm_minor_heap_size(wsize);
 
+  if (wsize > caml_minor_heap_max_wsz) wsize = caml_minor_heap_max_wsz;
```
I'm wondering whether this branch would be taken.

kayceesrk replied:
Alternatively, I'm wondering whether this line could be turned into

```c
CAMLassert (wsize <= caml_minor_heap_max_wsize);
```

gasche replied:
Good question, I am not sure. It may be that we now always update the max minor heap size in the case where this branch would have been taken in the past. This would be a nice simplification, thanks!

sabine replied:
I think so, yes, changing to the assert.

kayceesrk commented on runtime/domain.c:
```diff
@@ -398,8 +438,25 @@ int caml_reallocate_minor_heap(asize_t wsize)
+                   (uintnat) domain_state->young_start);
+}
+
+int caml_allocate_minor_heap(asize_t wsize) {
```
```suggestion
Caml_inline int allocate_minor_heap(asize_t wsize) {
```

kayceesrk commented on runtime/domain.c:
```diff
@@ -390,6 +420,16 @@ int caml_reallocate_minor_heap(asize_t wsize)
       && domain_state->young_end <= (value*)domain_self->minor_heap_area_end));
+}
+
+void caml_free_minor_heap() {
```
```suggestion
Caml_inline void free_minor_heap() {
```

gasche replied:
Is there a reason to `inline` this? It does not sound performance-critical. Would `extern` be enough?

kayceesrk replied:
You are right. Perhaps `inline` is an overkill. `static void` since the function is not used outside this file? Same for my other suggestions.

gasche replied:
@sabine changed this to `static`, which I think is better here.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+  }
+}
+
+void caml_update_minor_heap_max(uintnat minor_heap_wsz) {
```
```suggestion
void caml_update_minor_heap_max(uintnat requested_wsz) {
```

for clarity; avoids confusion with `Caml_state->minor_heap_wsz`.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
-  if (caml_mem_round_up_pages(Bsize_wsize(Minor_heap_max))
-          != Bsize_wsize(Minor_heap_max))
-    caml_fatal_error("Minor_heap_max misconfigured for this platform");
+  if (caml_mem_round_up_pages(Bsize_wsize(caml_minor_heap_max_wsz))
```
This should be an assertion rather than a check that runs always.

gasche replied:
Actually we need to be careful here, because there is no such rounding check in `gc_ctrl.c` which can change the value. With the PR as it currently stands (an assertion instead of a runtime check), the user can pass non-rounded values in `Gc.set` and the code runs.

If I understand correctly, this check is here for performance reasons, not for correctness: we are worried about negative performance impact of two minor heap areas sharing the same page? (Or is it an artifact of an old implementation using bit tricks that we don't need anymore?)

I'm not sure whether the correct change is:
- remove the check altogether
- maintain the invariant that `caml_minor_heap_max_wsz` is always rounded up (when it is initially defined, and when it is changed)

sabine replied:
Calls to `mmap` must be page-aligned: "If addr is not NULL, then the kernel takes it as a hint about where to place the mapping; on Linux, the kernel will pick a nearby page boundary" (https://man7.org/linux/man-pages/man2/mmap.2.html). The same holds for committing memory via `VirtualAlloc`: "If the memory is already reserved and is being committed, the address is rounded down to the next page boundary." (https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc)

I think that rounding up is the better developer experience, so I'll make the change.

sabine replied:
Turns out there was only one location where we did not call `caml_norm_minor_heap_size`. Fixed.

kayceesrk commented on runtime/domain.c:
```diff
@@ -423,9 +480,18 @@ int caml_reallocate_minor_heap(asize_t wsize)
+int caml_reallocate_minor_heap(asize_t wsize)
+{
+  caml_free_minor_heap();
+
```
Minor: delete line.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
-  void* heaps_base;
+/* minor heap initialization and resizing */
 
+static uintnat get_minor_heap_reservation_size() {
```
```suggestion
Caml_inline uintnat get_minor_heap_reservation_size() {
```

kayceesrk replied:
The two places where `get_minor_heap_reservation_size` is used, the result is immediately passed to `reserve_minor_heaps`. Best to push `get_minor_heap_reservation_size` code into `reserve_minor_heaps` and delete that function.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+          * Max_domains);
+}
+
+static void reserve_minor_heaps(uintnat size) {
```
```suggestion
Caml_inline void reserve_minor_heaps(uintnat size) {
```

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+    struct dom_internal* dom = &all_domains[i];
+
+    uintnat domain_minor_heap_area = caml_minor_heaps_base +
+      (uintnat)Bsize_wsize(caml_minor_heap_max_wsz) * (uintnat)i;
```
It is odd that the function takes `size` as an argument but uses `caml_minor_heap_max_wsz` here. This is correct only under the assumption that `size == Bsize_wsize(caml_minor_heap_max_wsz) * Max_domains`. 

See my comment elsewhere about pushing `get_minor_heap_reservation_size` into this function. That way, `reserve_minor_heaps` will not take any arguments.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+
+    dom->minor_heap_area = domain_minor_heap_area;
+    dom->minor_heap_area_end =
+         domain_minor_heap_area + Bsize_wsize(caml_minor_heap_max_wsz);
```
Minor: define a local variable `minor_heap_max_bsz`.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+    dom->minor_heap_area = dom->minor_heap_area_end = 0;
+  }
+
+  size = caml_minor_heaps_end - caml_minor_heaps_base;
```
I assume
```c
CAMLassert (Bsize_wsize(caml_minor_heap_max_wsz) * Max_domains == size)
```
holds here.

gasche replied:
I think so! In the past we would update `caml_minor_heap_wsz` to the new/desired value *before* calling this function, so we needed to be careful to use the "real size" and not the "desired size" here. But the current version is more robust (no slightly inconsistent state), and I think the proposed assert holds. (It could be falsified again with more flexible heap-area-reservation schemes, for example if we used an exponential-resizing strategy instead of requesting exactly the new needed size each time.)

kayceesrk replied:
Would the assertion still be useful to document the current behaviour?

gasche replied:
Yes of course; I am planning to take a global look at all comments and update the PR. (But probably later in the day or week.)

sabine replied:
I added the assert.

kayceesrk commented on runtime/domain.c:
```diff
@@ -335,36 +354,36 @@ void caml_domain_set_name(char *name)
     [caml_minor_heaps_base]
   and
-    [caml_minor_heaps_base_end}.
+    [caml_minor_heaps_base_end].
```
```suggestion
    [caml_minor_heaps_end].
```

kayceesrk commented on runtime/domain.c:
```diff
@@ -335,36 +354,36 @@ void caml_domain_set_name(char *name)
-    [domain_self->minor_heap_base]
+- Each domain gets a reserved section of this block
+  of size [caml_minor_heap_max_wsz], whose boundaries are stored as
+    [domain_self->minor_heap_area]
```
Generally, in this section, the terminology of the pointer that points to the end of an area is consistent but not the one that points to the start. See

1. `caml_minor_heaps_base`, `caml_minor_heaps_end`
2. `minor_heap_area`, `minor_heap_area_end`
3. `young_start`, `young_end`

Given that `young_start` has a long precedent, I would be in favour of the following renaming:

1. `minor_heap_area` -> `minor_heap_area_start`
2. `caml_minor_heaps_base` -> `caml_minor_heaps_start`.

sabine replied:
Sounds reasonable to me, the naming was inconsistent indeed.

kayceesrk commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
+  caml_mem_unmap((void *) caml_minor_heaps_base, size);
+}
+
+static void caml_stw_resize_minor_heap_reservation(caml_domain_state* domain,
```
```suggestion
static void stw_resize_minor_heap_reservation(caml_domain_state* domain,
```

sabine commented on runtime/domain.c:
```diff
@@ -630,28 +699,146 @@ CAMLexport void caml_reset_domain_lock(void)
-  void* heaps_base;
+/* minor heap initialization and resizing */
 
+static uintnat get_minor_heap_reservation_size() {
```
Done

gasche commented on runtime/domain.c:
```diff
@@ -701,39 +700,38 @@ CAMLexport void caml_reset_domain_lock(void)
-          * Max_domains);
-}
+  CAMLassert (caml_mem_round_up_pages(Bsize_wsize(caml_minor_heap_max_wsz))
+          != Bsize_wsize(caml_minor_heap_max_wsz));
```
Minor: this should be `==` here.

sabine committed:
Reserve only caml_minor_heap_max_wsz * Max_domains for the minor heap

When a domain uses Gc.set to increase its own minor heap size to be larger
than the current reservation, we reserve a new memory area for
all minor heaps. This is done inside a STW section:

1. every domain frees its own heap,
2. inside a global barrier, one domain unreserves the old minor heap
area, then reserves the new minor heap area. This domain also adjusts
the minor heap area pointers of all domains while the other domains
wait for the end of the barrier,
3. every domain allocates its own heap with size `Caml_state->minor_heap_wsz`.

Co-authored-by: Gabriel Scherer <gabriel.scherer@gmail.com>

sabine commented:
Reason for the latest force-pushes: I saw that we forgot to call `reset_minor_tables` after the heap allocation in the STW section and was rushing to add that, then there were some minor naming fixes.

Now, I think this is in a shape that we could merge it. I have no idea why the build timed out on the `tests/memory-model/'forbidden.ml'` test case, locally it works fine.

I'm ready to make a follow up PR to change the semantics of `Gc.set` to modify the size of all minor heaps (which is possibly the most obvious API to the user and also the simplest one to implement) - from the current semantics which only modifies the calling domain's minor heap size. This addresses the FIXME and results in some simplifications. The branch is at https://github.com/sabine/ocaml/tree/gc_set_changes_minor_heap_size_of_all_domains.

kayceesrk reviewed (approved):
The PR looks good to me. Hoping that we can quickly follow-up with a PR that fixes the double reservation for the domain requesting `Gc.set`: https://github.com/ocaml/ocaml/pull/11082#discussion_r829301114.

xavierleroy commented:
What's the status on this PR?  Should we merge based on @kayceesrk 's approval, or are we expecting more changes?  E.g. @ctk21's remark "I was surprised that a new STW section was introduced rather than just adding the functionality onto the end of the existing minor heap collection; you always have to empty the minor heaps to resize" was never answered, as far as I can see.

ctk21 commented:
For the avoidance of doubt, I just queried if the design could be simplified by reusing the existing STW section. However there isn't anything wrong with the way it has been done with an additional STW section.

sabine commented:
Good point! The reason for a separate STW section is that we wanted to keep the existing minor heap collection STW section as simple as possible. It is already a nontrivial part of the codebase. We felt that a separate STW section for the resizing results in better separation of concerns. Also, we do not risk introducing any performace degradation to the existing minor collection.

Note: The minor-heap-resizing STW section reuses code of the existing minor collection STW section, as we need to collect before we resize.

gasche commented:
@ctk21 actually I'm not sure what you mean. I can think of two interpretations:
- Maybe the code, at runtime, performs two STW sections in sequence, and this could be simplified conceptually (and made more efficient by removing more synchronization) by packing the two operations in a single STW section. But I don't think this is the case here.
- Or maybe the idea is to modify the *code* of the minor gc to also deal with heap resizing, but it's not clear to me what the benefits would be. It would make the minor-gc entry point more complex, and we would have to ensure that our modifications (when not used) are performance-neutral. Currently we already share code by calling the function `caml_empty_minor_heap_no_major_slice_from_stw` from our STW section, as the major GC is already doing.

gasche commented on runtime/domain.c:
```diff
@@ -648,28 +717,147 @@ CAMLexport void caml_reset_domain_lock(void)
+    caml_fatal_error("Fatal error: No memory for minor heap");
+  }
+
+  caml_reset_minor_tables(Caml_state->minor_tables);
```
Why are we calling `caml_reset_minor_tables` here? I don't remember why we added this and I can't convince myself that it is actually necessary -- otherwise we could add a comment to explain why.

gasche replied:
The minor tables should be empty at this point -- they get cleared by `caml_empty_minor_heap_domain_clear` which is called by `caml_stw_empty_minor_heap_no_major_slice` (both functions live in minor_gc.c).

I think the `reset` may have been added to mimick the logic in `set_minor_heap_size`, but the two functions are doing something rather different: `set_minor_heap_size` is changing the live minor heap size, and there it might make set to reset GC tables to go back to default capacity (when the call is made to substantially *shrink* the minor heap) if we expect a substantially lighter workflow in the future?

sabine replied:
In trunk, `reset_minor_tables` is called after the reallocation of the minor heap, at the end of  `caml_set_minor_heap_size`
https://github.com/ocaml/ocaml/blob/0c249195f84fbc0ef475c1334594b53aaf5a3de7/runtime/minor_gc.c#L122-L135

My impression was that we forgot to add that, due to oversight. Unless there is a good reason why we do not need to reset the minor tables after we free, re-reserve, and allocate all the minor heaps.

sabine replied:
> The minor tables should be empty at this point -- they get cleared by caml_empty_minor_heap_domain_clear which is called by caml_stw_empty_minor_heap_no_major_slice (both functions live in minor_gc.c).

Ok, you are right, these tables are already cleared and we do not need to call `reset_minor_tables`. I'm removing the call.

gasche replied:
Thanks! Marking this resolved.

ctk21 commented:
My thinking was the second ("modify the code of the minor gc"); the argument for this would be that it keeps all the minor heap code in one place and we would just check to see if the resize code path is needed at the point we hit the minor GC collection barrier (that is the point that all domains signal they have collected their minor heap). An argument against is that the STW minor GC collection function is big and hairy.

It seems like you considered this in your design and decided what you are doing is better. That's fine!

xavierleroy commented on Changes:
```diff
@@ -64,6 +64,16 @@ Working version
   (Enguerrand Decorne, report by Jon Ludlam,
   review by Tom Kelly, KC Sivaramakrishnan and Gabriel Scherer)
 
+- #11082 The runtime now reserves only minor_heap_wsz words of
+  memory per domain for the minor heap area (down from 2GB per domain).
```
What is `minor_heap_wsz` ?  Is the reader of the Changes file supposed to know?

Also, `down from 2GB per domain` is inexact since commit aff58350d.

xavierleroy commented on Changes:
```diff
@@ -64,6 +64,16 @@ Working version
+  the current domain.
+  Tools like AFL or Valgrind can now be used on compiled programs.
+  (Sabine Schmaltz, Gabriel Scherer, KC Sivaramakrishnan, review by
+   Enguerrand Decorne, Armaël Guéneau, Jacques-Henri Jourdan, ??)
```
`??` needs updating

xavierleroy commented on testsuite/tests/parallel/domain_parallel_spawn_burn_gc_set.ml:
```diff
@@ -0,0 +1,63 @@
+      };
+      set_gc (l + 1)
+
+let test_size =
+  try int_of_string (Sys.getenv "OCAML_TEST_SIZE")
+  with Not_found | Failure _ -> 0
+
+let (list_size, num_domains, niters) =
+  if test_size >= 2 then (14, 25, 20) else (13, 12, 5)
```
I appreciate the parameterization by `OCAML_TEST_SIZE`.  Still, this test starts much more domains than available processor cores (2, 4, or 8 depending on `OCAML_TEST_SIZE`).  Is this wise?  Is this needed?

sabine replied:
Very good point, indeed a smaller test size will do! Fixed.

xavierleroy commented on Changes:
```diff
@@ -64,6 +64,16 @@ Working version
+  reservation will un-reserve and re-reserve the minor heap area
+  of all domains. Gc.set changes the minor heap size only for
+  the current domain.
+  Tools like AFL or Valgrind can now be used on compiled programs.
```
Again, AFL and Valgrind were somehow usable since commit aff58350d.  What did not work is programs like Coq and CompCert that request big minor heaps.

xavierleroy reviewed (commented):
I'm still unsure what needs to be done to get this PR merged.  At any rate here are silly comments on the test and the Changes entry.  I'm not sure we need a Changes entry, but if we have one it should explain the changes w.r.t. 4.14, not w.r.t. Multicore OCaml.

sabine commented:
Yes, a Changes entry should not be needed because none of the possible issues arising from large heap reservation were ever exposed to the user. All we're doing here is to fix the memory reservation/allocation strategy in such a way that it does what the user already expects (i.e. minimal amount of reservation and allocation).

Looking at 4.14 as the reference point for the Changes entry is the correct thing to do. Thank you for reminding, it's something we need to take into account for all the multicore-related PRs.

ctk21 commented on tools/eventlog_metadata.in:
```diff
@@ -133,7 +133,8 @@ enum gc_phase : uint16_t {
     "minor_finalizers_admin",
-    "domain/condition_wait"
+    "domain/condition_wait",
+    "resize minor heap reservation"
```
This new field is the only one with the space character `' '` in it. I don't know if this will cause pain elsewhere, but it feels odd not to match the style of the other fields. 
```suggestion
    "resize_minor_heap_reservation"
```

abbysmal replied:
Indeed, it should to be `resize_minor_heap_reservation`

abbysmal replied:
Actually it should be `domain/resize_minor_heap_reservation`.
Because the event is called `EV_DOMAIN_RESIZE_MINOR_HEAP_RESERVATION`.

sabine replied:
Thank you! It's done

abbysmal reviewed (approved):
I took a last look at this PR and besides one final nitpick to the `eventlog_metadata.in` file I think things looks good here.

kayceesrk commented:
Thanks, @sabine and @gasche for the work, and all the reviewers for their careful reviews.
